src/common/getType.ts

```diff
@@ -400,57 +404,82 @@ function getOpCodeType(
     case "text_concat": {
       expectType(textType(), textType());
       const [t1, t2] = types as [TextType, TextType];
-      return textType(t1.capacity + t2.capacity);
+      return textType(
+        getArithmeticType("add", t1.codepointLength, t2.codepointLength),
+        t1.isAscii && t2.isAscii
+      );
     }
     case "repeat": {
       expectType(textType(), integerType(0));
       const [t, i] = types as [TextType, IntegerType];
-      return textType(mul(BigInt(t.capacity), i.high));
+      return textType(
+        getArithmeticType("mul", t.codepointLength, i),
+        t.isAscii
+      );
     }
     case "text_contains":
       expectType(textType(), textType());
       return booleanType;
-    case "text_find":
+    case "text_codepoint_find":
+    case "text_byte_find":
       expectType(textType(), textType());
-      return integerType(-1, (types[0] as TextType).capacity - 1);
+      return integerType(
+        -1,
+        mul(
+          sub((types[0] as TextType).codepointLength.high, 1n),
+          expr.op === "text_byte_find" && !(types[0] as TextType).isAscii
+            ? 4n
+            : 1n
+        )
+      );
     case "text_split":
       expectType(textType(), textType());
       return listType(types[0]);
-    case "text_get_char":
+    case "text_get_byte":
+    case "text_get_codepoint":
       expectType(textType(), integerType(0));
-      return textType(1);
+      return textType(integerType(1, 1), (types[0] as TextType).isAscii);
```

Comment:
This suggests the potential of keeping track of byte length separate from codepoint length: `text_get_byte` would alway returns a 1-byte string, while `text_get_codepoint` gives a 1-codepoint string. Not in the scope of this PR, maybe just something to keep in mind.